planner/core/find_best_task.go

```diff
@@ -534,6 +534,10 @@ func (ds *DataSource) getIndexMergeCandidate(path *util.AccessPath) *candidatePa
 func (ds *DataSource) skylinePruning(prop *property.PhysicalProperty) []*candidatePath {
 	candidates := make([]*candidatePath, 0, 4)
 	for _, path := range ds.possibleAccessPaths {
+		// skip tiflash if the statement is for stale read until tiflash support stale read
+		if ds.ctx.GetSessionVars().StmtCtx.IsStaleness && path.StoreType == kv.TiFlash {
```

Comment:
I would suggest you inject the filter into `getPossibleAccessPaths`, which is more essential and is executed while constructing logic plan.

`skylinePruning` is executed while constructing physical plan and it only applies to `DataSource`. It seems that `LogicalUnionAll` in `rule_partition_processor` will not execute `skylinePruning`, should verify this case.